### PR TITLE
feat: Get locale from cookie 

### PR DESCRIPTION
**Pull request by ioanna0**
## What does this change?

- Adds cookie module 
- Change locale logic to get the country code from newly [GU_geo_country](https://github.com/guardian/fastly-edge-cache/pull/906) cookie
- Return overridden geolocation if it exist to allow RR team test their banners in different countries
- Call API as fallback when cookie is not present and set its value to cookie
- Frontend PR: https://github.com/guardian/frontend/pull/23739

## Why?
Faster & more accurate geolocation

### CODE REVIEW

File: src/getLocale.ts
Comment (mxdvl) [inline src/getLocale.ts:30]: I think we’d want to do a few things here to make this more robust:

1. check whether `isValidCountryCode`
2. move this inside `try / catch`, otherwise we risk having out of sync cookies, memoized variable and local storage.

Comment (ioanna0) [reply src/getLocale.ts]: > 1. check whether isValidCountryCode

I added the check although I think this check is really unnecessary since fastly already returns a valid [Two-letter ISO 3166-1 alpha-2 country code for the country. ](https://developer.fastly.com/reference/vcl/variables/geolocation/client-geo-country-code/)


> 2. move this inside try / catch, otherwise we risk having out of sync cookies, memoized variable and local storage.

Why is that? this new cookie will not be cached at all since it will always be returned from fastly on page load time. if this is not the case (e.g localhost or some domain that doesn't go through fastly) the cookie will `null` and then we will continue with that we were doing before - fetch country code from geolocation endpoint and cache it.

Comment (mxdvl) [reply src/getLocale.ts]: If we’re memoizing the `locale`, so that it can be accessed without needing an extra check (cookie or local storage), it’s worth leveraging that benefit.

I think that the reading from the cookie should come in place of a `fetch` call to the API, but not as a short-circuit upstream. Otherwise, it would be impossible to force a user into a locale on prod, as the cookie override it every time.

Trying to diagram it, I find `A` chain is easier to reason about than `B`.

```
A. let locale? → localStorage? → cookie? → API!

B. cookie? → let locale? → localStorage? → API!
```

Comment (ioanna0) [reply src/getLocale.ts]: Agreed with memoizing it 👍
I've changed it to:
```
C. locale? → cookie? (If yes set to locale)→ localStorage? (If yes set to locale) → API! (Set response to local & localStorage)
```
Does that makes sense?
In prod we shouldn't touch localStorage or API I think.

**Review by mxdvl — DISMISSED** [review]
🗺️

**Comment by ioanna0** [comment]
@guardian/client-side-infra can you please review this? 🙏 I think I need 1 more approval before merging

File: src/getLocale.ts
Comment (ioanna0) [reply src/getLocale.ts]: I had a re-think of this and I think keeping localStorage in local could lead to confusion. It would be better to keep the same storage as production - i.e cookie, so thinking of changing to that?

`locale? → cookie? (If yes set to locale)→ API! (Set response to cookie)`

File: src/cookies.README.md
Comment (mxdvl) [inline src/cookies.README.md:26]: I think you can get rid of this 😉 

```suggestion
Type: `string`
```

File: src/cookies.ts
Comment (mxdvl) [inline src/cookies.ts]:
```suggestion
export const setSessionCookie = (name: string, value: string): void => {
```

**Review by mxdvl — APPROVED** [review]
LGTM, with some small comments.

I’ll let @guardian/client-side-infra give the last +1

**Comment by ioanna0** [comment]
> ioanna0 dismissed mxdvl’s stale review via e3a8e17 27 seconds ago
^ That's automated 🙄

**Review by mxdvl — DISMISSED** [review]

**Comment by ioanna0** [comment]
@gtrufitt and/or @guardian/client-side-infra can you review please?

**Review by arelra — DISMISSED** [review]

File: src/cookies.README.md
Comment (sndrs) [inline src/cookies.README.md:8]: i think this example import is wrong?

File: src/cookies.test.ts
Comment (sndrs) [inline src/cookies.test.ts:29]: could this be a simpler test value? 😝

File: src/cookies.test.ts
Comment (sndrs) [inline src/cookies.test.ts:35]: is this testing anything that the example above isn't testing?

File: src/cookies.test.ts
Comment (sndrs) [inline src/cookies.test.ts:42]: might be useful to `expect` this to be null before setting it?

File: src/cookies.test.ts
Comment (sndrs) [inline src/cookies.test.ts:47]:
```suggestion
	it('should be able to clear a list of cookies', () => {
```

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:69]: how do you feel about exporting `removeCookie` and removing this method? users can map over a list of cookies if they need, and the export name is then super clear about what is happening?

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:53]: for completeness, do you want to add a `setCookie` method?

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:61]: getting cookies is a blocking action – is this something we could safely memoize do you think?

File: src/getLocale.ts
Comment (sndrs) [inline src/getLocale.ts:7]: 🤩

File: src/getLocale.ts
Comment (sndrs) [inline src/getLocale.ts:12]: i know it's not part of this PR but it would be good to cache this regex rather than recreating it every time this is called

File: src/index.ts
Comment (sndrs) [inline src/index.ts:9]: i think this is missing some exports?

File: src/getLocale.ts
Comment (sndrs) [inline src/getLocale.ts:27]:
```suggestion
	// return overridden geo from localStorage, used for changing geo only for development purposes
```

**Review by sndrs — CHANGES_REQUESTED** [review]
this looks really good, so much better than hitting an endpoint! 

couple of things to look at and a couple of questions too

File: src/cookies.README.md
Comment (ioanna0) [reply src/cookies.README.md]: Yes totally wrong! 😳 fixed now

File: src/cookies.test.ts
Comment (ioanna0) [reply src/cookies.test.ts]: lol indeed, I got it from `frontend` I think a much simpler cookie value is also testing enough of the logic changed it to a trimmed value

File: src/cookies.test.ts
Comment (ioanna0) [reply src/cookies.test.ts]: Nope! the more advanced cookie above is covering what we need.

File: src/cookies.test.ts
Comment (ioanna0) [reply src/cookies.test.ts]: this now tests the `remove` method instead

File: src/index.ts
Comment (ioanna0) [reply src/index.ts]: Exported all of them now! Ready to be used as a replacement module of [frontend cookies.js](https://github.com/guardian/frontend/blob/main/static/src/javascripts/lib/cookies.js)

File: src/cookies.ts
Comment (ioanna0) [reply src/cookies.ts:87]: i think that's a great idea that would be mostly applicable for cookies that don't change their values for the duration of the session. i don't know if we can assume this would be the case for all cookies? 🤔
maybe we can do it for only specific cookies that we def know they won't change such as country code?

Comment (mxdvl) [reply src/cookies.ts:87]: Maybe adding an extra optional parameter like `memoizeValue = true` would be enough? Is it worth adding JSDoc comments to the lib to describe this behaviour?

```js
/** Reads a cookie, and optionally memoizes its value.
 * @param {string} name - the cookie’s name.
 * @param {boolean} memoizeValue - if set to `false`, calls will not memoize the cookie’s value.
 */
export const getCookie = (name: string, memoizeValue = true): string | null => {
  /* … */
}

```

Comment (ioanna0) [reply src/cookies.ts:123]: Good idea, I added the boolean in the `setCookie` and `setSessionCookie` instead

File: src/getLocale.ts
Comment (ioanna0) [reply src/getLocale.ts:12]: which one is the expensive part? should we memoize the results of calling `test`, or do you mean just the pattern bit?

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:57]: have you considered using a params object here? it would enable to you write (and later read) uses like this:

```js
setCookie({ name: "a", value: "b", isCrossSubdomain: true })
setCookie({ name: "a", value: "b", daysToLive: 1, shouldMemoize: true })
```

here's a [example in typescript](https://www.typescriptlang.org/play?#code/MYewdgzgLgBBCmUDCIQGsCW8YF4YAoBvAKBhjAEMBbeAGlJgDcKAbAVzoYBMKBPCACogAMhkacyGCEgBOICBADKbAEZcQVChjC4YAM1YJ6ZCAAsQbFlwCy8KiAwAvbHgMsjxAL4AuGCTKUNL7QMtoA5gDcDMzs8MFQoWCR3HyCImLwAPy+YGxUKvAyUZLScgrKahpaYNkwKqgs8BRgxXDmljZ2Ds619SCNzVGeAJS4AHx+DKCQ-fAAdCwgYUTk1HRMrBy0MDz8QqLi21Ky8kqq6pra22YWVrb2TtgjQ1HECMiomPArgXEwAEQUf4wZ5vRAodBYH5rXyA-7bGIcWEqYGg94Qr7QoIAoEIzZ-f4o7a7NIHP4ARhBwyi6M+UMIq2xcLxsWR8J2qX2GV85KOpVOFQu1V8CQ4VJp4Lp3wZv1huI2rIBRJgxzKZ0qlzAIpkYtBQA)

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:83]:
```suggestion
 * Set a session cookieookieookie. If it's been memoized it won't retrieve it again.
```

File: src/cookies.test.ts
Comment (sndrs) [inline src/cookies.test.ts:47]: i think not all the ways this could be called are captured here? e.g. 

```js
setCookie('a', 'b', , true)
setCookie('a', 'b', 1, , true)
```

**Review by sndrs — CHANGES_REQUESTED** [review]
couple of test cases missing, also wondered what you thought about using a params rather than a list of args?

File: src/getLocale.ts
Comment (sndrs) [reply src/getLocale.ts]: yeah that's one, saves recreating a regex for `/^[A-Z]{2}$/` each time it runs – thanks :)

File: src/cookies.ts
Comment (ioanna0) [reply src/cookies.ts]: I just changed it again sorry! i think what Max suggested initially to memoize on `getCookie` might be better because we want to memoize on fetch time. We cannot assume that a setCookie has happened in the same session as a get i think? And I changed the setCookie to override the memoized value if it exists.
But great idea with params object, could be useful if we are back to 5 params again

Comment (ioanna0) [reply src/cookies.ts]: https://github.com/guardian/libs/pull/162/commits/ef4c8daa1279f93ef6886209d649a12f9d307f33

Comment (sndrs) [reply src/cookies.ts]: i wonder if memoizing cookie reading is really just a premature optimisation that only adds complexity? (sorry i know i brought it up!)

File: src/cookies.README.md
Comment (ioanna0) [reply src/cookies.README.md]: changed to: `setCookie({name, value, daysToLive?, isCrossSubdomain?})``

File: src/cookies.ts
Comment (ioanna0) [reply src/cookies.ts]: maybe but it could also be really useful! hopefully it will be used by consumers. since it's only optional functionality maybe there's no harm leaving it in? it can also be refactored if needs be

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:48]: i think these are out of date now?

File: src/cookies.README.md
Comment (sndrs) [inline src/cookies.README.md:18]:
```suggestion
-   [`setCookie({name, value, daysToLive?, isCrossSubdomain?})`](#setCookie)
```

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:91]: sorry to be a pain, but do you think it would be clearer if all cookie methods took a consisten type of payload – i.e. a params object like `setCookie`?

Comment (sndrs) [reply src/cookies.ts:91]: i feel like i would expect that having done 
```js
setCookie({ name: 'a', value: 'b' })
``` 
i'd expect to be able to do something like this too: 
```js
setSessionCookie({ name: 'a', value: 'b' })
getCookie({ name: 'a', shouldMemoize: false })
```

**Review by sndrs — CHANGES_REQUESTED** [review]
some docs outdated now i think, and a question around signature consistency

File: src/cookies.ts
Comment (ioanna0) [reply src/cookies.ts]: Yes, though so but at that time i though it was simpler to pass params for 1 or 2 values rather than config obj. Is better to be consistent though across the whole module and have refactored it now across all those methods

File: src/cookies.ts
Comment (ioanna0) [reply src/cookies.ts]: how does that look like?
```
 * @param details Details about the cookie.
 * @param details.name - the cookie’s name.
 * @param details.value - the cookie’s value.
 * @param details.daysToLive - optional expiry date will be calculated based on the daysToLive
 * @param details.isCrossSubdomain - specify if it's a cross subdomain cookie, default false
```

File: src/cookies.ts
Comment (sndrs) [inline src/cookies.ts:111]: this is the only one now not accepting an object – sorry, can it?

Comment (sndrs) [reply src/cookies.ts:111]: also, i don't think `currentDomainOnly` is documented?

**Review by sndrs — CHANGES_REQUESTED** [review]
nearly – just these last 2 things then all good

File: src/cookies.test.ts
Comment (mxdvl) [inline src/cookies.test.ts:124]: Don't you want to set a specific number of times it's expected to be called?

```suggestion
		expect(spy).toHaveBeenCalledTimes(1);
```

File: src/cookies.ts
Comment (mxdvl) [inline src/cookies.ts:21]: Is it worth extracting this Regex to a constant, as we've done for the `validCountryCode`?

**Review by sndrs — DISMISSED** [review]
great! thanks so much for your patience @ioanna0. we discussed this offline, but with a lib like this the API gets set so early on that (hopefully) there's a lot of value in persevering at this stage, so thank you for bearing with all the change requests :)

excellent addition 🎉

File: src/cookies.ts
Comment (sndrs) [reply src/cookies.ts:21]: good spot

File: src/cookies.test.ts
Comment (ioanna0) [reply src/cookies.test.ts:114]: thanks, good catch! I completely missed the `not` and now is making total sense, is calling it just once ! 🙂

Comment (mxdvl) [reply src/cookies.test.ts:114]: Yes, exactly, It was **not** _not_ doing what you wanted! 😃

**Review by sndrs — APPROVED** [review]
lovely stuff :)

**Comment by sndrs** [comment]
AOK to merge?

**Comment by ioanna0** [comment]
Yes!

**Pull request merged**
